packages/gatsby/src/utils/websocket-manager.ts

```diff
@@ -176,22 +151,39 @@ export class WebsocketManager {
       }
 
       const getDataForPath = async (path: string): Promise<void> => {
-        if (!this.pageResults.has(path)) {
+        let pageData = this.pageResults.get(path)
+        if (!pageData) {
           try {
-            const result = await getCachedPageData(path)
+            pageData = await getPageData(path)
 
-            this.pageResults.set(path, result)
+            this.pageResults.set(path, pageData)
           } catch (err) {
             console.log(err.message)
-
             return
           }
         }
 
+        const staticQueryHashes = pageData.result?.staticQueryHashes ?? []
+        await Promise.all(
+          staticQueryHashes.map(async queryId => {
+            let staticQueryResult = this.staticQueryResults.get(queryId)
```

Comment:
This effectively has the same race condition as the page data cache (because they are resolved in parallel in `promise.all`). In every step the `staticQueryResults` cache is queried before any element in this loop (at least) has a chance to finish and populate it.

Reply:
Sure, if 6 clients connect at the same time, we might overfetch by 6. We can do this in a follow up but doesn't seem a big concern to me.